### ISSUE TITLE
type inference on createStructuredCachedSelector?

### ISSUE THREAD

**Issue by parkerault** [toomuchdesign/re-reselect#103]
### Do you want to request a _feature_ or report a _bug_?

Maybe a bug?

### What is the current behaviour?

Type error when calling `createStructuredCachedSelector` without type parameters:

Typescript fails to infer the type parameters:
```typescript
const state = { a: 1, b: "asdf " };
type State = typeof state;

const mySelectorA = (state: State) => state.a;
const mySelectorB = (state: State) => state.b;

const structuredSelector = createStructuredCachedSelector({
  x: mySelectorA,
  y: mySelectorB,
})((s: State) => "asdf");
/*
No overload matches this call.
  Overload 2 of 2, '(selectors: { x: ParametricSelector<unknown, unknown, number>; y: ParametricSelector<unknown, unknown, string>; }): OutputParametricCachedSelector<unknown, unknown, { x: number; y: string; }, (...args: (string | number)[]) => { x: number; y: string; }, ParametricSelector<unknown, unknown, string | number>[]>', gave the following error.
    Type '(state: { a: number; b: string; }) => number' is not assignable to type 'ParametricSelector<unknown, unknown, number>'.
      Types of parameters 'state' and 'state' are incompatible.
        Type 'unknown' is not assignable to type '{ a: number; b: string; }'.
  Overload 2 of 2, '(selectors: { x: ParametricSelector<unknown, unknown, number>; y: ParametricSelector<unknown, unknown, string>; }): OutputParametricCachedSelector<unknown, unknown, { x: number; y: string; }, (...args: (string | number)[]) => { x: number; y: string; }, ParametricSelector<unknown, unknown, string | number>[]>', gave the following error.
    Type '(state: { a: number; b: string; }) => string' is not assignable to type 'ParametricSelector<unknown, unknown, string>'.
      Types of parameters 'state' and 'state' are incompatible.
        Type 'unknown' is not assignable to type '{ a: number; b: string; }'.
*/
```

Specifying the type parameters explicitly works:
```typescript
const state = { a: 1, b: "asdf " };
type State = typeof state;

const mySelectorA = (state: State) => state.a;
const mySelectorB = (state: State) => state.b;

const structuredSelector = createStructuredCachedSelector<
  State,
  { x: ReturnType<typeof mySelectorA>; y: ReturnType<typeof mySelectorB> }
>({
  x: mySelectorA,
  y: mySelectorB,
})((s: State) => "asdf");
```

### What is the expected behaviour?

Are the type parameters supposed to be inferred or explicitly defined? `createCachedSelector` infers the type parameters so I'm not sure if this is a bug or not, and the documentation doesn't clarify this.

**Comment by toomuchdesign** [comment]
I suspect it's a typing issue originated in `reselect` and ported here :)
https://github.com/reduxjs/reselect/issues/428

**Comment by toomuchdesign** [comment]
Optimistically shipped with v4.0.0.
